src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1563,21 +1563,25 @@ private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancell
         return newCompilationState;
     }
 
-    private (ImmutableDictionary<ProjectId, ProjectState>, ImmutableSegmentedDictionary<ProjectId, ICompilationTracker>) ComputeFrozenSnapshotMaps(CancellationToken cancellationToken)
+    private (ImmutableArray<ProjectState>, ImmutableSegmentedDictionary<ProjectId, ICompilationTracker>) ComputeFrozenSnapshotMaps(CancellationToken cancellationToken)
     {
         // Loop until we have calculated the maps against a set of compilation trackers that hasn't changed during our calculation.
         while (true)
         {
             var originalProjectIdToTrackerMap = _projectIdToTrackerMap;
-            var newIdToProjectStateMapBuilder = this.SolutionState.ProjectStates.ToBuilder();
+            using var _ = ArrayBuilder<ProjectState>.GetInstance(out var newProjectStatesBuilder);
             var newIdToTrackerMapBuilder = originalProjectIdToTrackerMap.ToBuilder();
+            var projectStateChanged = false;
 
             // Used to track any new compilation trackers created in the loop. This is done to avoid allocations
             // from individually adding to the _projectIdToTrackerMap ImmutableSegmentedDictionary .
             var updatedIdToTrackerMapBuilder = originalProjectIdToTrackerMap.ToBuilder();
 
-            foreach (var projectId in this.SolutionState.ProjectIds)
+            newProjectStatesBuilder.AddRange(this.SolutionState.ProjectStates);
```

Comment:
Is this better than just calling the .ToBuilder() method on the original array? I would have imagined this might enumerate more, but maybe not.

Reply:
ImmutableArray.Builder is a class that requires an allocation. ArrayBuilder.AddRange uses the pooled ImmutableArray.Builder to do the work for it.